### ISSUE TITLE
Cloud: start_action prevented by highstate

### ISSUE THREAD

**Issue by syphernl** [saltstack/salt#23763]
Iḿ currently testing out the VMware cloud driver. For this I have created a VM template which gets cloned. There are however some system-specific tasks to deal with prior to it being provisioned (e.g. growing partitions, changing NIC naming, installing "base" updates).

For this I have created a state that takes care of kickstarting everything. The Salt-cloud configuration includes a `start_action`:

```
start_action: state.sls virtual.provision
```

The debug trail shows it is being executed. However, I believe because the salt-master has (pre)signed the key the pillarized schedule kicks in and starts provisioning immediately.

This results in the `start_action` not being executed at all.

```
[INFO    ] Executing state.sls virtual.provision on the salt-minion
[DEBUG   ] Using sudo to run command sudo salt-call state.sls virtual.provision
[DEBUG   ] SSH command: "ssh -t -t -oStrictHostKeyChecking=no -oUserKnownHostsFile=/dev/null -oControlPath=none -oPasswordAuthentication=no -oChallengeResponseAuthentication=no -oPubkeyAuthentication=yes -oKbdInteractiveAuthentication=no -i /root/.ssh/id_rsa -p 22 user@172.xx.xx.xx 'sudo salt-call state.sls virtual.provision'"
[DEBUG   ] Child Forked! PID: 9036  STDOUT_FD: 19  STDERR_FD: 21
[DEBUG   ] Terminal Command: /bin/sh -c ssh -t -t -oStrictHostKeyChecking=no -oUserKnownHostsFile=/dev/null -oControlPath=none -oPasswordAuthentication=no -oChallengeResponseAuthentication=no -oPubkeyAuthentication=yes -oKbdInteractiveAuthentication=no -i /root/.ssh/id_rsa -p 22 user@xx.xx.xx 'sudo salt-call state.sls virtual.provision'
Warning: Permanently added '172.xx.xx.xx' (ECDSA) to the list of known hosts.
[DEBUG   ] Warning: Permanently added '172.xx.xx.xx' (ECDSA) to the list of known hosts.
Connection to 172.xx.xx.xx closed.
[DEBUG   ] Connection to 172.xx.xx.xx closed.
local:
    Data failed to compile:
----------
    The function "state.highstate" is running as PID 2141 and was started at 2015, May 15 12:50:05.570274 with jid 20150515125005570274
[DEBUG   ] local:
    Data failed to compile:
----------
    The function "state.highstate" is running as PID 2141 and was started at 2015, May 15 12:50:05.570274 with jid 20150515125005570274
```

The `virtual.provision` state does not contain anything about highstates (e.g. applying the env/minion specific tasks in `top.sls`), so where does this particular error originate from?

Version specs:

```
                  Salt: 2015.5.0
                Python: 2.7.6 (default, Mar 22 2014, 22:59:56)
                Jinja2: 2.7.2
              M2Crypto: 0.21.1
        msgpack-python: 0.3.0
          msgpack-pure: Not Installed
              pycrypto: 2.6.1
               libnacl: Not Installed
                PyYAML: 3.10
                 ioflo: Not Installed
                 PyZMQ: 14.0.1
                  RAET: Not Installed
                   ZMQ: 4.0.4
                  Mako: 0.9.1
       Apache Libcloud: 0.14.1
 Debian source package: 2015.5.0+ds-1trusty1
```

**Comment by jfindlay** [comment]
@syphernl, thanks for the report.

**Comment by techhat** [comment]
Thanks @syphernl. This feature is still marked as experimental, as it has not ever been known to function correctly:

http://docs.saltstack.com/en/latest/topics/cloud/deploy.html#post-deploy-commands

It is far more reliable to use a reactor to respond to `salt/cloud/*/created` events. IIRC, `start_action` was added before reactors were, but was essentially trying to do the same thing.

**Comment by jfindlay** [comment]
@syphernl, this feature of salt-cloud, as @techhat indicates, is superseded by better functionality in the reactor system.

**Comment by syphernl** [comment]
I've been using the reactors a while back, wasn't particularly convinced about it.

Wouldn't the reactor fire after (or at the same time) of the automated highstate? I'm looking for something to fire as soon as the VM is up..

**Comment by techhat** [comment]
With `salt/cloud/*/created`, the reactor would fire once Salt Cloud has reported that it is finished. This is likely to be at least several seconds after startup states.

`start_action` looks for the `minion_start` event to appear on the bus:

https://github.com/saltstack/salt/blob/develop/salt/utils/cloud.py#L2027

This event is deprecated in favor of the namespaced `salt/minion/*/start` event. But basically, if you configure a `start_action`, Salt Cloud will create a new subprocess to look for it. Unfortunately, the subprocess doesn't play well with SSH, which causes the unreliability. This was particularly problematic with paramiko, which was one of the reasons why we ripped it out a couple of years ago, but it doesn't like shelling out to SSH much either.

The reactor, being a separate process entirely, doesn't encounter the same issues with SSH, and so is much more reliable. Whatever shortcomings you may have encountered with the reactor system, I would say it's still generally far more reliable than `start_action`.

**Comment by syphernl** [comment]
Hmm, seems like this doesn't work though. 

I had to set `start_action` to a value that doesn't do much (`test.ping`). 
If it is not defined it results in executing a highstate, even though the highstate is already being initiated by the scheduler. This causes error output during provisioning

I have created a reactor for `salt/cloud/*/created` but it looks like the automated highstate (pushed via schedules) is blocking execution:

```
[INFO    ] Got return from testvm5.example.com for job 20150519112706269768
[DEBUG   ] Sending event - data = {'fun_args': ['virtual.provision'], 'jid': '20150519112706269768', 'return': ['The function "state.highstate" is running as PID 2088 and was started at 2015, May 19 11:26:02.713216 with jid 20150519112602713216'], 'retcode': 1, 'success': True, 'cmd': '_return', '_stamp': '2015-05-19T09:27:06.453371', 'fun': 'state.sls', 'id': 'testvm5.example.com', 'out': 'highstate'}
[DEBUG   ] Gathering reactors for tag salt/job/20150519112706269768/ret/testvm5.example.com
```

**Comment by techhat** [comment]
Am I safe to assume that the `start_action` that you're performing is a `highstate`? And that you have the minion itself performing a `highstate` using a Salt schedule?

**Comment by syphernl** [comment]
The `start_action` in the salt-cloud config is `test.ping`. For normal salt-operations this is not specifically configured.

The master has a pillar with the schedule which gets pushed to the minion upon start.

**Comment by aphor** [comment]
#25555 <-- I think this is another case where people want salt cloud to work like/with salt state.highstate.

https://github.com/saltstack/salt/issues/25555#issuecomment-153553006 <-- my suggestion to rework salt cloud backend using proxy minions would expose cloud APIs to states with pillar and grain data, requisites, and reactors on beacons or events.
